crypto/crmf/crmf_lib.c

```diff
@@ -0,0 +1,753 @@
+        goto err;
+    }
+    crlen = (size_t)len;
+
+    if (!OBJ_find_sigid_by_algs(&alg_nid, dgst, EVP_PKEY_id(pkey))) {
+        CRMFerr(CRMF_F_CRMF_POPOSIGNINGKEY_INIT,
+                CRMF_R_UNSUPPORTED_ALG_FOR_POPSIGNINGKEY);
+        goto err;
+    }
+    if (!OBJ_find_sigid_algs(alg_nid, &md_nid, NULL)
+            || (alg = EVP_get_digestbynid(md_nid)) == NULL) {
+        CRMFerr(CRMF_F_CRMF_POPOSIGNINGKEY_INIT,
+                CRMF_R_UNSUPPORTED_ALG_FOR_POPSIGNINGKEY);
+        goto err;
+    }
+    if (!X509_ALGOR_set0(ps->algorithmIdentifier, OBJ_nid2obj(alg_nid),
+                         V_ASN1_NULL, NULL)
+            || (ctx = EVP_MD_CTX_new()) == NULL
+            || EVP_DigestSignInit(ctx, NULL, alg, NULL, pkey) <= 0
+            || EVP_DigestSignUpdate(ctx, crder, crlen) <= 0
+            || EVP_DigestSignFinal(ctx, NULL, &siglen) <= 0) {
+        CRMFerr(CRMF_F_CRMF_POPOSIGNINGKEY_INIT, CRMF_R_ERROR);
+        goto err;
+    }
+    if ((sig = OPENSSL_malloc(siglen)) == NULL) {
+        CRMFerr(CRMF_F_CRMF_POPOSIGNINGKEY_INIT, ERR_R_MALLOC_FAILURE);
+        goto err;
+    }
+    if (EVP_DigestSignFinal(ctx, sig, &siglen) <= 0
+            || !ASN1_BIT_STRING_set(ps->signature, sig, siglen)) {
+        CRMFerr(CRMF_F_CRMF_POPOSIGNINGKEY_INIT, CRMF_R_ERROR);
+        goto err;
+    }
+    ret = 1;
+
+ err:
+    OPENSSL_free(crder);
+    EVP_MD_CTX_free(ctx);
+    OPENSSL_free(sig);
+    return ret;
+}
+
+
+int OSSL_CRMF_MSG_create_popo(OSSL_CRMF_MSG *crm, EVP_PKEY *pkey,
+                              int dgst, int ppmtd)
+{
+    OSSL_CRMF_POPO *pp = NULL;
+    ASN1_INTEGER *tag = NULL;
+
+    if (crm == NULL || (ppmtd == OSSL_CRMF_POPO_SIGNATURE && pkey == NULL)) {
+        CRMFerr(CRMF_F_OSSL_CRMF_MSG_CREATE_POPO, CRMF_R_NULL_ARGUMENT);
+        return 0;
+    }
+
+    if (ppmtd == OSSL_CRMF_POPO_NONE)
+        goto end;
+    if ((pp = OSSL_CRMF_POPO_new()) == NULL)
+        goto oom;
+    pp->type = ppmtd;
+
+    switch (ppmtd) {
+    case OSSL_CRMF_POPO_RAVERIFIED:
+        if ((pp->value.raVerified = ASN1_NULL_new()) == NULL)
+            goto oom;
+        break;
+
+    case OSSL_CRMF_POPO_SIGNATURE:
+        {
+            OSSL_CRMF_POPOSIGNINGKEY *ps = OSSL_CRMF_POPOSIGNINGKEY_new();
+            if (ps == NULL
+                    || !CRMF_poposigningkey_init(ps, crm->certReq, pkey, dgst)){
+                OSSL_CRMF_POPOSIGNINGKEY_free(ps);
+                goto err;
+            }
+            pp->value.signature = ps;
+        }
+        break;
+
+    case OSSL_CRMF_POPO_KEYENC:
+        if ((pp->value.keyEncipherment = OSSL_CRMF_POPOPRIVKEY_new()) == NULL)
+            goto oom;
+        pp->value.keyEncipherment->type =
+            OSSL_CRMF_POPOPRIVKEY_SUBSEQUENTMESSAGE;
+        tag = ASN1_INTEGER_new();
+        if (tag == NULL
+                || !ASN1_INTEGER_set(tag, OSSL_CRMF_SUBSEQUENTMESSAGE_ENCRCERT))
+            goto oom;
+        pp->value.keyEncipherment->value.subsequentMessage = tag;
```

Comment:
resource leak.
please move this assignment before the if. If you like combine it with pp->...subsequentMessage = tag = ...new();

Reply:
.. or explicitly free tag like the ps above.

Reply:
Good spotting! I've moved the assignment (even one more further up, to keep the two assignments to `pp->value.keyEncipherment->type` adjacent).
Combining the two assignments would have been nice but would have gone beyond the 80-char line limit.